test/e2e/kb/version_upgrade_test.go

```diff
@@ -0,0 +1,120 @@
+	"github.com/elastic/cloud-on-k8s/pkg/controller/kibana/label"
+	"github.com/elastic/cloud-on-k8s/pkg/utils/k8s"
+	"github.com/elastic/cloud-on-k8s/test/e2e/test"
+	"github.com/elastic/cloud-on-k8s/test/e2e/test/elasticsearch"
+	"github.com/elastic/cloud-on-k8s/test/e2e/test/kibana"
+	"github.com/stretchr/testify/assert"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func TestVersionUpgrade(t *testing.T) {
+	name := "test-version-upgrade"
+	esBuilder := elasticsearch.NewBuilder(name).
+		WithESMasterDataNodes(1, elasticsearch.DefaultResources).
+		WithVersion("7.4.1")
+
+	kbBuilder := kibana.NewBuilder(name).
+		WithElasticsearchRef(esBuilder.Ref()).
+		WithNodeCount(3).
+		WithVersion("7.4.0")
+
+	ns := client.InNamespace(kbBuilder.Kibana.Namespace)
+	matchLabels := client.MatchingLabels(map[string]string{
+		common.TypeLabelName:      label.Type,
+		label.KibanaNameLabelName: kbBuilder.Kibana.Name,
+	})
+
+	var observations []int
+	w := test.NewWatcher("watch pods readiness", 1*time.Second,
+		func(k *test.K8sClient, t *testing.T) {
+			if pods, err := k.GetPods(ns, matchLabels); err != nil {
+				t.Logf("got error: %v", err)
+			} else {
+				ready := 0
+				for _, pod := range pods {
+					if k8s.IsPodReady(pod) {
+						ready++
+					}
+				}
+				observations = append(observations, ready)
+			}
+		},
+		func(k *test.K8sClient, t *testing.T) {
+			assert.Contains(t, observations, 0)
+		})
+
+	test.RunMutationsWhileWatching(
+		t,
+		[]test.Builder{esBuilder, kbBuilder},
+		[]test.Builder{esBuilder, kbBuilder.WithVersion("7.4.1").WithMutatedFrom(&kbBuilder)},
+		[]test.Watcher{w},
+	)
+}
+
+func TestRespecAndVersionUpgrade(t *testing.T) {
+	name := "test-upgrade-and-respec"
+	esBuilder := elasticsearch.NewBuilder(name).
+		WithESMasterDataNodes(1, elasticsearch.DefaultResources).
+		WithVersion("7.4.1")
+
+	kbBuilder1 := kibana.NewBuilder(name).
+		WithElasticsearchRef(esBuilder.Ref()).
+		WithNodeCount(3).
+		WithVersion("7.4.0")
+
+	kbBuilder2 := kbBuilder1.WithMutatedFrom(&kbBuilder1).WithVersion("7.4.1")
+	kbBuilder3 := kbBuilder2.WithMutatedFrom(&kbBuilder2).WithLabel("some", "label")
+
+	ns := client.InNamespace(kbBuilder1.Kibana.Namespace)
+	matchLabels := client.MatchingLabels(map[string]string{
+		common.TypeLabelName:      label.Type,
+		label.KibanaNameLabelName: kbBuilder1.Kibana.Name,
+	})
+
+	// checks whether after temporary downtime kibana will be available the rest of the time
+	var hadZero, shouldHaveNonZero, failed bool
+	w := test.NewWatcher("watch pods readiness", 1*time.Second,
+		func(k *test.K8sClient, t *testing.T) {
+			pods, err := k.GetPods(ns, matchLabels)
+			if err != nil {
+				t.Logf("got error: %v", err)
+			}
+
+			ready := 0
+			for _, pod := range pods {
+				if k8s.IsPodReady(pod) {
+					ready++
+				}
+			}
+
+			hadZero = hadZero || ready == 0
+			if hadZero && ready > 0 {
+				shouldHaveNonZero = true
+			}
+
+			if shouldHaveNonZero && ready == 0 {
+				failed = true
+			}
```

Comment:
same comment re checking downtime vs. checking multiple versions running in parallel.

Reply:
This one is meant to check whether we don't do recreate twice, ie. whether after first upgrade, the label add change will result in a rolling upgrade. Checking for pods version doesn't achieve confirming that the rollout is rolling upgrade. 

In other words, if we hit a race condition/bug during second upgrade (label add) and the operator will use recreate again, the versions will be consistent, but since they were killed and started at the same time the pods won't be ready. Which is what I'd like to detect here.

Reply:
Chatted offline, we will have both.